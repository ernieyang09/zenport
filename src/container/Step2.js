// @flow
import React, { useState, useCallback } from 'react';
import { SelectPicker, Button } from 'rsuite';
import { useMappedState ,useDispatch} from 'redux-react-hook';

import {
  clickPrevStep,
  clickNextStep,
} from 'store/modules/ui';

import type {
  UIAction,
} from 'store/modules/ui';

import {
  formSelector,
  setRestaurant,
} from 'store/modules/form';

import type {
  FormAction,
} from 'store/modules/form';

import {
  dishesSelector,
} from 'store/modules/dishes';

import type {
  Dish
} from 'store/modules/dishes';

import {
  ButtonPanel,
  ErrMsg,
} from 'styles/pages';





const Step2 = () => {
  const { meal, restaurant } = useMappedState(formSelector);

  const mapState = useCallback((state) => {
    const dishes: Array<Dish> = dishesSelector(state);
    const restaurants: Array<string> = dishes
      .filter(d => d.availableMeals.some(m => m === meal))
      .map(d => d.restaurant)
    return [...new Set(restaurants)]
  }, [meal]);

  const restaurants = useMappedState(mapState);

  const [select, setSelect] = useState(restaurant);
  const [valid, setValid] = useState({ restaurant: true });

  const validate = () => {
    const restaurant = (select !== null);
    setValid({ restaurant });
    return restaurant;
  }
  

  const dispatch: ((FormAction | UIAction) => void) = useDispatch();


  return (
    <div>
      <div>Please select a restaurant</div>
      <div>
        <SelectPicker
          data={restaurants.map(r => ({ label: r, value: r }))}
          searchable={false}
          value={select}
          onChange={(val: string) => { setSelect(val); }}
        />
      </div>
      {
        !valid.restaurant && <ErrMsg>Please select a restaurant</ErrMsg>
      }
      <ButtonPanel>
        <Button
          onClick={()=> {
            dispatch(setRestaurant({ restaurant: select }));
            dispatch(clickPrevStep());
          }}
        >
          PREV
        </Button>
        <Button
          data-test-id='next'
          onClick={()=> { 
            if (!validate()) {
              return
            }
            dispatch(setRestaurant({ restaurant: select }));
            dispatch(clickNextStep());
          }}
        >
          NEXT
        </Button>
      </ButtonPanel>
    </div>
  )
}

export default Step2;
