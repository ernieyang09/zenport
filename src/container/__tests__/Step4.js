import React from 'react';
import {StoreContext} from 'redux-react-hook';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';

import Step2 from '../Step2';

import {
  clickNextStep,
} from 'store/modules/ui';

import {
  setRestaurant,
} from 'store/modules/form';

const mockStore = configureStore([]);

const fakeStore = {
  dishes: [
    {
      id: 1,
      name: 'Chicken Burger',
      restaurant: 'Mc Donalds',
      availableMeals: [
        'lunch',
        'dinner'
      ]
    },
    {
      id: 2,
      name: 'Ham Burger',
      restaurant: 'Mc Donalds',
      availableMeals: [
        'lunch',
        'dinner'
      ]
    },
    {
      id: 3,
      name: 'Cheese Burger',
      restaurant: 'Mc Donalds',
      availableMeals: [
        'lunch',
        'dinner'
      ]
    },
    {
      id: 4,
      name: 'Fries',
      restaurant: 'Mc Donalds',
      availableMeals: [
        'lunch',
        'dinner'
      ]
    },
    {
      id: 5,
      name: 'Egg Muffin',
      restaurant: 'Mc Donalds',
      availableMeals: [
        'breakfast'
      ]
    },
    {
      id: 6,
      name: 'Burrito',
      restaurant: 'Taco Bell',
      availableMeals: [
        'lunch',
        'dinner'
      ]
    },
    {
      id: 7,
      name: 'Tacos',
      restaurant: 'Taco Bell',
      availableMeals: [
        'lunch',
        'dinner'
      ]
    },
    {
      id: 8,
      name: 'Quesadilla',
      restaurant: 'Taco Bell',
      availableMeals: [
        'lunch',
        'dinner'
      ]
    },
    {
      id: 9,
      name: 'Steak',
      restaurant: 'BBQ Hut',
      availableMeals: [
        'dinner'
      ]
    },
    {
      id: 10,
      name: 'Yakitori',
      restaurant: 'BBQ Hut',
      availableMeals: [
        'dinner'
      ]
    },
    {
      id: 11,
      name: 'Nankotsu',
      restaurant: 'BBQ Hut',
      availableMeals: [
        'dinner'
      ]
    },
    {
      id: 12,
      name: 'Piman',
      restaurant: 'BBQ Hut',
      availableMeals: [
        'dinner'
      ]
    },
    {
      id: 13,
      name: 'Vegan Bento',
      restaurant: 'Vege Deli',
      availableMeals: [
        'lunch'
      ]
    },
    {
      id: 14,
      name: 'Coleslaw Sandwich',
      restaurant: 'Vege Deli',
      availableMeals: [
        'breakfast'
      ]
    },
    {
      id: 15,
      name: 'Grilled Sandwich',
      restaurant: 'Vege Deli',
      availableMeals: [
        'breakfast'
      ]
    },
    {
      id: 16,
      name: 'Veg. Salad',
      restaurant: 'Vege Deli',
      availableMeals: [
        'lunch',
        'dinner'
      ]
    },
    {
      id: 17,
      name: 'Fruit Salad',
      restaurant: 'Vege Deli',
      availableMeals: [
        'lunch',
        'dinner'
      ]
    },
    {
      id: 18,
      name: 'Corn Soup',
      restaurant: 'Vege Deli',
      availableMeals: [
        'lunch',
        'dinner'
      ]
    },
    {
      id: 19,
      name: 'Tomato Soup',
      restaurant: 'Vege Deli',
      availableMeals: [
        'lunch',
        'dinner'
      ]
    },
    {
      id: 20,
      name: 'Minestrone Soup',
      restaurant: 'Vege Deli',
      availableMeals: [
        'lunch',
        'dinner'
      ]
    },
    {
      id: 21,
      name: 'Pepperoni Pizza',
      restaurant: 'Pizzeria',
      availableMeals: [
        'lunch',
        'dinner'
      ]
    },
    {
      id: 22,
      name: 'Pepperoni Pizza',
      restaurant: 'Pizzeria',
      availableMeals: [
        'lunch',
        'dinner'
      ]
    },
    {
      id: 23,
      name: 'Hawaiian Pizza',
      restaurant: 'Pizzeria',
      availableMeals: [
        'lunch',
        'dinner'
      ]
    },
    {
      id: 24,
      name: 'Seafood Pizza',
      restaurant: 'Pizzeria',
      availableMeals: [
        'lunch',
        'dinner'
      ]
    },
    {
      id: 25,
      name: 'Deep Dish Pizza',
      restaurant: 'Pizzeria',
      availableMeals: [
        'dinner'
      ]
    },
    {
      id: 26,
      name: 'Chow Mein',
      restaurant: 'Panda Express',
      availableMeals: [
        'lunch',
        'dinner'
      ]
    },
    {
      id: 27,
      name: 'Mapo Tofu',
      restaurant: 'Panda Express',
      availableMeals: [
        'lunch',
        'dinner'
      ]
    },
    {
      id: 28,
      name: 'Kung Pao',
      restaurant: 'Panda Express',
      availableMeals: [
        'lunch',
        'dinner'
      ]
    },
    {
      id: 29,
      name: 'Wontons',
      restaurant: 'Panda Express',
      availableMeals: [
        'lunch',
        'dinner'
      ]
    },
    {
      id: 30,
      name: 'Garlic Bread',
      restaurant: 'Olive Garden',
      availableMeals: [
        'breakfast',
        'lunch',
        'dinner'
      ]
    },
    {
      id: 31,
      name: 'Ravioli',
      restaurant: 'Olive Garden',
      availableMeals: [
        'lunch',
        'dinner'
      ]
    },
    {
      id: 32,
      name: 'Rigatoni Spaghetti',
      restaurant: 'Olive Garden',
      availableMeals: [
        'lunch',
        'dinner'
      ]
    },
    {
      id: 33,
      name: 'Fettucine Pasta',
      restaurant: 'Olive Garden',
      availableMeals: [
        'lunch',
        'dinner'
      ]
    }
  ],
  ui: {
    step: 1
  },
  form: {
    meal: 'dinner',
    number: 4,
    restaurant: 'Mc Donalds',
    orders: [
      {
        key: 0,
        dish: null,
        quantity: 1
      }
    ]
  }
};

describe('Step4', () => {
  let store;
  let component;
  beforeEach(() => {
    store = mockStore(fakeStore);
    store.dispatch = jest.fn();

    component = renderer.create(
      <StoreContext.Provider value={store}>
        <Step2 />
      </StoreContext.Provider>
    );
  });

  

  it('should render with given state from Redux store', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('dispatch next button click', () => {
    renderer.act(()=> {
      component.root.findByProps({ 'data-test-id': "next" }).props.onClick();
      expect(store.dispatch).toHaveBeenCalledWith(setRestaurant({ restaurant: 'Mc Donalds' }));
      expect(store.dispatch).toHaveBeenCalledWith(clickNextStep());
      expect(store.dispatch).toHaveBeenCalledTimes(2);
    })
  });
});
