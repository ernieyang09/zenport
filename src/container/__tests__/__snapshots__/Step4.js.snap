// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Step4 should render with given state from Redux store 1`] = `
<div>
  <div>
    Please select a restaurant
  </div>
  <div>
    <div
      className="rs-picker-select rs-picker-default rs-picker-toggle-wrapper rs-picker-placement-bottom-start rs-picker-has-value rs-picker-cleanable"
      onClick={[Function]}
      role="menu"
      tabIndex={-1}
    >
      <a
        className="rs-picker-toggle"
        onBlur={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="combobox"
        tabIndex={0}
      >
        <span
          className="rs-picker-toggle-value"
        >
          Mc Donalds
        </span>
        <span
          className="rs-picker-toggle-clean"
          onClick={[Function]}
          role="button"
          tabIndex={-1}
        >
          âœ•
        </span>
        <span
          className="rs-picker-toggle-caret"
        />
        <span
          className="rs-ripple-pond"
        >
          <span
            className="rs-ripple"
            style={Object {}}
          />
        </span>
      </a>
    </div>
  </div>
  <div
    className="css-10urh76"
  >
    <button
      className="rs-btn rs-btn-default"
      onClick={[Function]}
      type="button"
    >
      PREV
      <span
        className="rs-ripple-pond"
      >
        <span
          className="rs-ripple"
          style={Object {}}
        />
      </span>
    </button>
    <button
      className="rs-btn rs-btn-default"
      data-test-id="next"
      onClick={[Function]}
      type="button"
    >
      NEXT
      <span
        className="rs-ripple-pond"
      >
        <span
          className="rs-ripple"
          style={Object {}}
        />
      </span>
    </button>
  </div>
</div>
`;
